sphinx/transforms/i18n.py

```diff
@@ -342,6 +342,14 @@ class Locale(SphinxTransform):
     default_priority = 20
 
     def apply(self, **kwargs: Any) -> None:
+        # TODO/FIXME: If this document is marked :notranslate:, it might
+        # be better to skip translation here. Some paragraphs might
+        # still get translated in rare cases, e.g., if gettext_compact=True
+        # and there is an identical paragraph in a document that is not
+        # marked :notranslate:. This might be tricky, since self.env.metadata
+        # is only populated in the doctree-read event, which runs after
+        # this transform.
```

Comment:
Oh dear. I just realized that this will also make the `translation_progress` variable wrong. It should not account for `:notranslate:` documents.

@AA-Turner Can you help here? Should I make the environment metadata collector run earlier somehow? Should I traverse the document tree directly to search for metadata in `is_translatable_document`?

Reply:
Correct me if I'm wrong but your issue is:

- You want to know whether the document should be translated or not.
- However this information is gathered by the collector which at the point when the transformation is executed is not yet called.
- So you end up translating everything even if it was marked `:notranslate:` (but was not yet detected).
- When you count the number of mistranslations, then you are also accounting those that were marked as `:notranslate:` since they were gathered during the `Locale` transform.

I don't know if running the metadata collector before is fine or not since it's triggered by the 'doctree-read' event. 

Maybe we could add a dedicated `MetadataCollectorTransform` which does the same but only for `:notranslate:` and run that transformation just before `Locale` by adding the information to the *doctree* itself and not to the environment?

Reply:
Sorry for the long hiatus. Yes, you've understood the problem correctly.

> Maybe we could add a dedicated MetadataCollectorTransform which does the same but only for :notranslate: and run that transformation just before Locale by adding the information to the doctree itself and not to the environment?

I don't know whether special-casing `:notranslate:` is the proper fix. I imagine that another solution is to just run the metadata collector earlier than `doctree-read`, but this might puzzle extension transforms that add metadata to the doctree in order to have it collected.

Reply:
Yes this is what I fear hence the special casing. We are already doing special casing for some stuff so I think it's fine but we might want to think of another solution. I don't have time for that now but maybe others have an idea?